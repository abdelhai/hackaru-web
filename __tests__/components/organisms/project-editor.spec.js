import { Store } from 'vuex-mock-store';
import Factory from '@/__tests__/__setups__/factory';
import ProjectEditor from '@/components/organisms/project-editor';
import ProjectList from '@/components/organisms/project-list';

describe('ProjectEditor', () => {
  let factory;
  let wrapper;

  const $store = new Store({});

  beforeEach(() => {
    $store.reset();
    factory = new Factory(ProjectEditor, {
      mocks: { $store }
    });
  });

  it('render correctly', () => {
    expect(factory.shallow().element).toMatchSnapshot();
  });

  describe('when submit and id is not defined', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({ name: 'Development', color: '#ff0' });
      wrapper.find('form').trigger('submit.prevent');
    });

    it('dispatch projects/addProject', () => {
      expect($store.dispatch).toHaveBeenCalledWith('projects/addProject', {
        name: 'Development',
        color: '#ff0'
      });
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'project',
        'addProject'
      );
    });

    it('emit pop', () => {
      expect(wrapper.emitted('pop')[0]).toEqual([{ component: ProjectList }]);
    });
  });

  describe('when submit and id is not defined but failed', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.find('form').trigger('submit.prevent');
    });

    it('does not emit pop', () => {
      expect(wrapper.emitted('pop')).toBeFalsy();
    });
  });

  describe('when submit and id is defined', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({ id: 1, name: 'Development', color: '#ff0' });
      wrapper.find('form').trigger('submit.prevent');
    });

    it('dispatch projects/updateProject', () => {
      expect($store.dispatch).toHaveBeenCalledWith('projects/updateProject', {
        id: 1,
        name: 'Development',
        color: '#ff0'
      });
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'project',
        'updateProject'
      );
    });

    it('emit pop', () => {
      expect(wrapper.emitted('pop')[0]).toEqual([{ component: ProjectList }]);
    });
  });

  describe('when submit and id is defined but failed', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.setData({ id: 1 });
      wrapper.find('form').trigger('submit.prevent');
    });

    it('does not emit pop', () => {
      expect(wrapper.emitted('pop')).toBeFalsy();
    });
  });

  describe('when click delete button and confirm is true', () => {
    beforeEach(() => {
      global.confirm = () => true;
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({ id: 1, name: 'Development', color: '#ff0' });
      wrapper.find('.delete-button').vm.$emit('click');
    });

    it('dispatch projects/deleteProject', () => {
      expect($store.dispatch).toHaveBeenCalledWith('projects/deleteProject', 1);
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'project',
        'deleteProject'
      );
    });

    it('emit pop', () => {
      expect(wrapper.emitted('pop')[0]).toEqual([{ component: ProjectList }]);
    });
  });

  describe('when click delete button and confirm is false', () => {
    beforeEach(() => {
      global.confirm = () => false;
      wrapper = factory.shallow();
      wrapper.setData({ id: 1, name: 'Development', color: '#ff0' });
      wrapper.find('.delete-button').vm.$emit('click');
    });

    it('does not dispatch projects/deleteProject', () => {
      expect($store.dispatch).not.toHaveBeenCalledWith(
        'projects/deleteProject'
      );
    });

    it('does not emit pop', () => {
      expect(wrapper.emitted('pop')).toBeFalsy();
    });
  });

  describe('when click delete button but failed', () => {
    beforeEach(() => {
      global.confirm = () => true;
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.setData({ id: 1 });
      wrapper.find('.delete-button').vm.$emit('click');
    });

    it('does not emit pop', () => {
      expect(wrapper.emitted('pop')).toBeFalsy();
    });
  });
});
