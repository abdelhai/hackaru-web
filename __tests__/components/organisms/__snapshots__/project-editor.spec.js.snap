// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectEditor render correctly 1`] = `
<section>
  <modal-header-stub>
    <btn-stub
      aria-label="ariaLabels.back"
      class="has-icon"
      type="button"
    >
      <icon-stub
        class="is-large"
        name="chevron-left-icon"
      />
    </btn-stub>
     
    <h1>
      titles.add
    </h1>
  </modal-header-stub>
   
  <form>
    <modal-item-stub>
      <modal-label-stub>
        
        name
      
      </modal-label-stub>
       
      <input
        class="input inline"
        placeholder="name"
        type="text"
      />
    </modal-item-stub>
     
    <modal-item-stub>
      <modal-label-stub>
        
        color
      
      </modal-label-stub>
       
      <color-select-stub
        colors="#4ab8b8,#7da87b,#a1c45a,#7a4579,#f95959,#ea8a8a,#ff9f68,#facf5a,#616161,#8ea6b4"
        value="#cccfd9"
      />
    </modal-item-stub>
     
    <modal-footer-stub>
      <btn-stub
        class="is-rounded is-primary"
        type="submit"
      >
        
        add
      
      </btn-stub>
       
      <!---->
    </modal-footer-stub>
  </form>
</section>
`;
